subsys/bluetooth/host/classic/avctp.c

```diff
@@ -0,0 +1,151 @@
+/** @file
+ *  @brief Audio Video Control Transport Protocol
+ */
+
+/*
+ * Copyright (c) 2015-2016 Intel Corporation
+ * Copyright (C) 2024 Xiaomi Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <string.h>
+#include <strings.h>
+#include <errno.h>
+#include <zephyr/sys/atomic.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/sys/util.h>
+
+#include <zephyr/bluetooth/hci.h>
+#include <zephyr/bluetooth/bluetooth.h>
+#include <zephyr/bluetooth/l2cap.h>
+#include <zephyr/bluetooth/classic/avctp.h>
+
+#include "avctp_internal.h"
+#include "host/hci_core.h"
+#include "host/conn_internal.h"
+#include "l2cap_br_internal.h"
+
+#define LOG_LEVEL CONFIG_BT_AVCTP_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(bt_avctp);
+
+#define AVCTP_CHAN(_ch) CONTAINER_OF(_ch, struct bt_avctp, br_chan.chan)
+
+K_MUTEX_DEFINE(avctp_mutex);
+#define AVCTP_LOCK() k_mutex_lock(&avctp_mutex, K_FOREVER)
+#define AVCTP_UNLOCK() k_mutex_unlock(&avctp_mutex)
+
+static struct bt_avctp_event_cb *event_cb;
+
+static void avctp_l2cap_connected(struct bt_l2cap_chan *chan)
+{
+	struct bt_avctp *session;
+
+	if (!chan) {
+		LOG_ERR("Invalid AVCTP chan");
+		return;
+	}
+
+	session = AVCTP_CHAN(chan);
+	LOG_DBG("chan %p session %p", chan, session);
+
+	session->ops->connected(session);
+}
+
+static void avctp_l2cap_disconnected(struct bt_l2cap_chan *chan)
+{
+	struct bt_avctp *session = AVCTP_CHAN(chan);
+
+	LOG_DBG("chan %p session %p", chan, session);
+	session->br_chan.chan.conn = NULL;
+
+	session->ops->disconnected(session);
+}
+
+static void avctp_l2cap_encrypt_changed(struct bt_l2cap_chan *chan, uint8_t status)
+{
+	LOG_DBG("");
+}
+
+static int avctp_l2cap_recv(struct bt_l2cap_chan *chan, struct net_buf *buf)
+{
+	LOG_DBG("");
+
+	/* TODO */
+
+	return 0;
+}
+
+int bt_avctp_connect(struct bt_conn *conn, struct bt_avctp *session)
+{
+	static const struct bt_l2cap_chan_ops ops = {
+		.connected = avctp_l2cap_connected,
+		.disconnected = avctp_l2cap_disconnected,
+		.encrypt_change = avctp_l2cap_encrypt_changed,
+		.recv = avctp_l2cap_recv,
+	};
+
+	if (!session) {
+		return -EINVAL;
+	}
+
+	session->br_chan.rx.mtu = BT_L2CAP_RX_MTU;
+	session->br_chan.chan.ops = &ops;
+	session->br_chan.required_sec_level = BT_SECURITY_L2;
+
+	return bt_l2cap_chan_connect(conn, &session->br_chan.chan,
+		BT_L2CAP_PSM_AVCTP);
+}
+
+int bt_avctp_disconnect(struct bt_avctp *session)
+{
+	if (!session) {
+		return -EINVAL;
+	}
+
+	LOG_DBG("session %p", session);
+
+	return bt_l2cap_chan_disconnect(&session->br_chan.chan);
+}
+
+int bt_avctp_register(struct bt_avctp_event_cb *cb)
+{
+	LOG_DBG("");
+
+	if (event_cb) {
+		return -EALREADY;
+	}
+
+	event_cb = cb;
+
+	return 0;
+}
+
+static int avctp_l2cap_accept(struct bt_conn *conn, struct bt_l2cap_server *server,
+	struct bt_l2cap_chan **chan)
+{
+	/* TODO */
+
+	return 0;
```

Comment:
Please return an error code, such as `-ENOTSUP`, instead of `0` if the feature is unsupported.

Reply:
Would change it to -ENOTSUP.

The process of AVCTP packets are included from the next pull request.
BTW I think it is still fine if an L2CAP PDU is received correctly but nothing happens.

Reply:
I don't see any changes here. Please confirm.

Reply:
Done